lib/samples/navigate_route_with_rerouting/navigate_route_with_rerouting.dart

```diff
@@ -0,0 +1,600 @@
+// limitations under the License.
+//
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:arcgis_maps/arcgis_maps.dart';
+import 'package:arcgis_maps_sdk_flutter_samples/common/common.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_tts/flutter_tts.dart';
+import 'package:path_provider/path_provider.dart';
+
+class NavigateRouteWithRerouting extends StatefulWidget {
+  const NavigateRouteWithRerouting({super.key});
+
+  @override
+  State<NavigateRouteWithRerouting> createState() =>
+      _NavigateRouteWithReroutingState();
+}
+
+class _NavigateRouteWithReroutingState extends State<NavigateRouteWithRerouting>
+    with SampleStateSupport {
+  // Create a controller for the map view.
+  final _mapViewController = ArcGISMapView.createController();
+  // A flag for when the map view is ready and controls can be used.
+  var _ready = false;
+  // A text-to-speech plugin to provide voice guidance.
+  late FlutterTts _flutterTts;
+  // A RouteTask to solve the route.
+  late RouteTask _routeTask;
+  // A RouteTracker to track the route.
+  late RouteTracker _routeTracker;
+  // A RouteResult to store the route result.
+  late RouteResult _routeResult;
+  // Rerouting parameters to enable rerouting.
+  late ReroutingParameters _reroutingParameters;
+  // A SimulatedLocationDataSource to simulate the location data source.
+  SimulatedLocationDataSource? _simulatedLocationDataSource;
+  // Graphics to show progress on the route.
+  late Graphic _remainingRouteGraphic;
+  late Graphic _routeTraveledGraphic;
+  // A PolylineBuilder to store the traversed route.
+  var _traversedRouteBuilder =
+      PolylineBuilder(spatialReference: SpatialReference.wgs84);
+  // San Diego Convention Center.
+  final _startPoint = ArcGISPoint(
+    x: -117.160386727,
+    y: 32.706608,
+    spatialReference: SpatialReference.wgs84,
+  );
+  // RH Fleet Aerospace Museum.
+  final _endPoint = ArcGISPoint(
+    x: -117.146679,
+    y: 32.730351,
+    spatialReference: SpatialReference.wgs84,
+  );
+  // Indicate whether the route is being navigated.
+  var _isNavigating = false;
+  var _resetToNavigationMode = false;
+  var _reset = false;
+  var _setupDataAndInitNavigation = false;
+  // Variables to show the remaining distance, time, and next direction.
+  var _routeStatus = '';
+  var _remainingDistance = '';
+  var _remainingTime = '';
+  var _nextDirection = '';
+  // Future to download the geodatabase.
+  late Future<String> _geodatabasePathFuture;
+  // Future to download the simulated location data source.
+  late Future<SimulatedLocationDataSource> _simulatedLocationDataSourceFuture;
+  // Stream subscriptions
+  late StreamSubscription<VoiceGuidance> _voiceGuidanceSubscription;
+  late StreamSubscription<TrackingStatus> _trackingStatusSubscription;
+  late StreamSubscription<void> _rerouteStartedSubscription;
+  late StreamSubscription<void> _rerouteCompletedSubscription;
```

Comment:
Declaring these `late` is problematic because "dispose()" might get called before they are set, which will cause an exception and crash. For example, if the network downloads are slow and the user hits "Back" before all the "init" methods run. So these StreamSubscriptions should be declared without `late` and as nullable.